Testing.Rmd

```diff
@@ -275,140 +382,640 @@ All expectations have a similar structure:
+Here are a couple of the functions most useful when writing tests:
+
+* `local_options()` / `with_options()` (see above)
+* `local_envvar()` / `with_envvar()` for temporarily setting an environment variable
+    ```{r, eval = FALSE}
+    # from tibble, in tests/testthat/test-utils.R
+    test_that("is_rstudio()", {
+      expect_false(withr::with_envvar(c(RSTUDIO = NA), is_rstudio()))
+      expect_true(withr::with_envvar(c(RSTUDIO = 1), is_rstudio()))
+    })
+    ```
+* `local_tempfile()`, `local_tempdir()`, and `local_file()` for creating a
+  self-destructing file or directory
+    ```{r, eval = FALSE}
+    # from roxygen2, in tests/testthat/test-collate.R
+    test_that("can read from file name with utf-8 path", {
+      path <- withr::local_tempfile(
+        pattern = "Universit\u00e0-",
+        lines = c("#' @include foo.R", NULL)
+      )
+      expect_equal(find_includes(path), "foo.R")
+    })
+    ```
+
+There are many other functions in withr, so check there before you write your own.
+If nothing exists that meets your need, `withr::defer()` is the general way to schedule some action at the end of a test.[^on-exit]
+
+[^on-exit]: Base R's `on.exit()` is another alternative, but it requires more from you.
+You need to capture the original state and write the restoration code yourself.
+Also remember to do `on.exit(..., add = TRUE)` if there's *any* chance a second `on.exit()` call could be added in the test.
+You probably also want to default to `after = FALSE`.
+
+What if you need to do more than, e.g., set an option, before you can execute your expectations?
+We're going to make the controversial recommendation that you tolerate a fair amount of duplication in test code, i.e. you can relax some of your DRY ("don't repeat yourself") tendencies.
+
+Here's a toy example to make things concrete.
+
+```{r}
+test_that("multiplication works", {
+  useful_thing <- 3
+  expect_equal(2 * useful_thing, 6)
+})
+
+test_that("subtraction works", {
+  useful_thing <- 3
+  expect_equal(5 - useful_thing, 2)
+})
+```
+
+In real life, `useful_thing` is usually a more complicated object that somehow feels burdensome to instantiate.
+Notice how `useful_thing <- 3` appears in more than once place.
+Conventional wisdom says we should DRY this code out.
+It's tempting to just move `useful_thing`'s definition outside of the tests:
+
+```{r}
+useful_thing <- 3
+
+test_that("multiplication works", {
+  expect_equal(2 * useful_thing, 6)
+})
+
+test_that("subtraction works", {
+  expect_equal(5 - useful_thing, 2)
+})
+```
+
+This does sort of work, because when `useful_thing` is not found in the test-specific environment, the search continues in the parent environment, where `useful_thing` will often be found.
+When testthat executes an entire test file, `useful_thing` will get defined and made available for subsequent tests in that file.
+However, during interactive development, there is no gesture to systematically identify and execute such code.
```

Comment:
This is particularly frustrating when the test at line 573 is failing and all the setup code is at line 1.  Or worse when it depends on code smooshed between two other tests at some random location that made sense at the time you wrotei t.